omegaconf/basecontainer.py

```diff
@@ -299,8 +300,15 @@ def expand(node: Container) -> None:
                     node._set_value(type_)
 
         if src._is_missing() and not dest._is_missing():
-            if is_structured_config(get_ref_type(src)):
-                expand(src)
+            src_ref_type = get_ref_type(src)
+            if is_structured_config(src_ref_type):
+                assert src_ref_type is not None
+                # Replace `src` with a prototype of its corresponding structured config
+                # whose fields are all missing (to avoid overwriting fields in `dest`).
+                src_data = get_structured_config_data(src_ref_type)
+                for v in src_data.values():
+                    v._set_value(MISSING)
+                src = DictConfig(src_ref_type(**src_data), ref_type=src_ref_type)
```

Comment:
I think you shouldn't instantiate the dataclass here.
did you try passing `content=src_data` directly?

Reply:
Yes that was my first attempt, it failed with:
```
omegaconf.errors.ValidationError: Invalid type assigned : dict is not a subclass of User. value: {'name': '???', 'age': '???'}
```

Reply:
However, looking into it more closely, all tests pass with:
```python
src = DictConfig(src_data)
```
which worries me a bit because we lost the annotated type of `src` by doing so...

I wonder if we may be missing a `ref_type` change for `dest`, in the same spirit of the `if isinstance(dest_node, AnyNode):` that you added. More precisely, looking for instance at this test:

```python
    def test_merge_structured_onto_dict_nested(self, class_type: str) -> None:
        module: Any = import_module(class_type)
        c1 = OmegaConf.create({"user": {"name": 7}})
        c2 = OmegaConf.merge(c1, module.MissingUserField)
        assert c1 == {"user": {"name": 7}}
        # type of name becomes str
        assert c2 == {"user": {"name": "7", "age": "???"}}
```
Wouldn't we want to also add at the end:
```python
assert c2._get_node("user")._metadata.ref_type == module.User
```
(right now such an assert fails, because the left hand side is `Any`)

Reply:
Yeah, obviously this is an unusual way to instantiate the DictConfig.
Normally, the DictConfig itself is calling get_structured_config_data(), now we are doing it from the outside.

1. Yes, add the extra check to the test.
2. I think you can fix it by following your `src = DictConfig(src_data)` with a direct assignment of the ref_type and I think also the object type.

Reply:
All (hopefully) done in 3deee99

Reply:
Looks okay.